### PR TITLE
brings back the watch cache

### PR DESCRIPTION
**Pull request by p0lyn0mial**
## Summary
In addition, it provides coverage of `DynamicDiscoverySharedInformerFactory` to prove it works when the cache is on

## Related issue(s)

Fixes #

### CODE REVIEW

File: test/e2e/watchcache/watchcache_enabled_test.go [hunk 1]
```diff
@@ -202,3 +214,60 @@ func collectCacheHitsFor(ctx context.Context, t *testing.T, config *rest.Config,
 	}
 	return totalCacheHits, prefixCacheHit
 }
+
+const resyncPeriod = 10 * time.Hour
+const byWorkspace = "byWorkspace"
+
+func testDynamicDiscoverySharedInformerFactory(t *testing.T, cfg *rest.Config, expectedGVR schema.GroupVersionResource, expectedResName string, expectedClusterName logicalcluster.Name) {
+	t.Logf("Starting DynamicDiscoverySharedInformerFactory and checking if it has %v with name %v in cluster %v", expectedGVR.String(), expectedResName, expectedClusterName)
```
Comment (sttts) [inline test/e2e/watchcache/watchcache_enabled_test.go:222]: nit: move "Checking if ..." down above the `require.Eventually`. It's important to have a flow of `t.Logs` in e2e to understand where something breaks.

File: test/e2e/watchcache/watchcache_enabled_test.go [hunk 2]
```diff
@@ -202,3 +214,60 @@ func collectCacheHitsFor(ctx context.Context, t *testing.T, config *rest.Config,
...
+		metadataClusterClient.Cluster(logicalcluster.Wildcard),
+		func(obj interface{}) bool { return true }, 10*time.Second,
+	)
+	err = ddsif.AddIndexers(cache.Indexers{byWorkspace: indexByWorkspace})
+	require.NoError(t, err)
+
+	go kcpSharedInformerFactory.Start(context.Background().Done())
+	ddsif.Start(context.Background())
+
+	require.Eventually(t, func() bool {
```
Comment (sttts) [inline test/e2e/watchcache/watchcache_enabled_test.go:243]: try `framework.Eventually`. Allows you to print with deduplication.

File: test/e2e/watchcache/watchcache_enabled_test.go [hunk 3]
```diff
@@ -202,3 +214,60 @@ func collectCacheHitsFor(ctx context.Context, t *testing.T, config *rest.Config,
...
+		listers, notSynced := ddsif.Listers()
+		for _, ns := range notSynced {
+			klog.Infof("DynamicDiscoverySharedInformerFactory hasn't synced %v yet", ns.String())
+		}
+
+		for gvr, lister := range listers {
+			obj, err := lister.List(labels.Everything())
+			require.NoError(t, err)
+			if gvr.String() == expectedGVR.String() {
```
Comment (sttts) [inline test/e2e/watchcache/watchcache_enabled_test.go:252]: nit: `if ... { continue }` for less indention

**Comment by sttts** [comment]
cc @shawn-hurley @ncdc

File: test/e2e/watchcache/watchcache_enabled_test.go [hunk 4]
```diff
@@ -202,3 +214,70 @@ func collectCacheHitsFor(ctx context.Context, t *testing.T, config *rest.Config,
...
+	t.Log("Waiting for KCP Shared Informer Factory to sync caches")
+	kcpSharedInformerFactory.WaitForCacheSync(ctx.Done())
+	t.Log("Starting DynamicDiscoverySharedInformerFactory")
+	ddsif.Start(context.Background())
+
+	t.Logf("Checking if DynamicDiscoverySharedInformerFactory has %v with name %v in cluster %v", expectedGVR.String(), expectedResName, expectedClusterName)
+	framework.Eventually(t, func() (success bool, reason string) {
+		listers, notSynced := ddsif.Listers()
+		for _, ns := range notSynced {
+			klog.Infof("DynamicDiscoverySharedInformerFactory hasn't synced %v yet", ns.String())
```
Comment (stevekuznetsov) [inline test/e2e/watchcache/watchcache_enabled_test.go:255]: nit: t.Log

File: test/e2e/watchcache/watchcache_enabled_test.go [hunk 5]
```diff
@@ -202,3 +214,70 @@ func collectCacheHitsFor(ctx context.Context, t *testing.T, config *rest.Config,
...
+			}
+			for _, o := range obj {
+				u := o.(*unstructured.Unstructured)
+				if u.GetName() == expectedResName && u.GetClusterName() == expectedClusterName.String() {
+					return true, ""
+				}
+			}
+		}
+		return false, fmt.Sprintf("haven't found %v with name %v in %v cluster", expectedGVR, expectedResName, expectedClusterName)
+	}, wait.ForeverTestTimeout*2, time.Millisecond*100, "DynamicDiscoverySharedInformerFactory hasn't observed %v with name %v in %v cluster", expectedGVR, expectedResName, expectedClusterName)
```
Comment (stevekuznetsov) [inline test/e2e/watchcache/watchcache_enabled_test.go:272]: Why forever * 2?

Comment (p0lyn0mial) [reply test/e2e/watchcache/watchcache_enabled_test.go:272]: I think that `wait.ForeverTestTimeout` is not enough. I could increase the QPS and lower the poll timeout to speed up the test.

Comment (p0lyn0mial) [reply test/e2e/watchcache/watchcache_enabled_test.go]: done.

**Comment by sttts** [comment]
`FAIL	github.com/kcp-dev/kcp/test/e2e/watchcache	103.031s`

**Comment by p0lyn0mial** [comment]
this time a binding hasn't been completed in time
```
    apibinding.go:60: Creating APIBinding root:e2e-org-mmpjp:e2e-workspace-shpgn|root-e2e-org-mmpjp-e2e-workspace-5cswl
    apibinding.go:64: 
        	Error Trace:	apibinding.go:64
        	            				watchcache_enabled_test.go:130
        	Error:      	Condition never satisfied
        	Test:       	TestWatchCacheEnabledForAPIBindings
```

**Comment by ncdc** [comment]
I can still consistently get a failure in one of my quota tests with the watch cache on after 5-7 iterations. It made it through 20 iterations with the watch cache off. No clue yet why.

**Comment by sttts** [comment]
> I can still consistently get a failure in one of my quota tests with the watch cache on after 5-7 iterations.

@ncdc can you describe exactly how to run them? Have we seen them fail outside of your machine?

**Comment by ncdc** [comment]
1. `bin/test-server` (enable the watch cache first in options.go first)
2. `go test ./test/e2e/quota -run TestKubeQuotaCoreV1TypesFromBinding -v -p 1 -parallel 1 -failfast -count 20 -args --use-default-kcp-server`

**Comment by sttts** [comment]
Moving forward here. With @p0lyn0mial 's sleep example the issue in quota can also be triggered without watch cache. Looks like we are hunting a bug here, but watch cache is not the root cause. Would prefer to give watch cache more soak time instead.

/lgtm
/approve

**Comment by p0lyn0mial** [comment]
pushed a fix for `test/e2e/watchcache/watchcache_enabled_test.go:285:32: cannot use context.Background() (type context.Context) as type <-chan struct {} in argument to ddsif.Start`. I had to rebase and use `StartPolling` instead.

**Pull request merged**